src/Processors/QueryPlan/Optimizations/optimizeUseNormalProjection.cpp

```diff
@@ -197,73 +207,120 @@ std::optional<String> optimizeUseNormalProjections(Stack & stack, QueryPlan::Nod
             continue;
         }
 
         if (best_candidate == nullptr || candidate.sum_marks < best_candidate->sum_marks)
+            best_candidate = &candidate;
+
+        /// All parts has been filtered out; no need to analyze further projections.
+        if (parent_reading_marks == 0)
+            break;
+    }
+
+    if (!best_candidate)
+        return {};
+
+    /// Identify projections selected as the best candidates and update their stat descriptions with appropriate logging
+    for (const auto & candidate : candidates)
+    {
+        if (&candidate == best_candidate)
         {
-            LOG_DEBUG(
-                logger,
-                "Projection {} is selected as current best candidate with {} marks to read, while original table needs to scan {} marks",
+            chassert(candidate.stat);
+            chassert(candidate.stat->description.empty());
+            candidate.stat->description = fmt::format(
+                "Projection {} is selected as the best with {} marks to read, while the original table requires scanning {} marks",
                 candidate.projection->name,
                 candidate.sum_marks,
-                ordinary_reading_marks);
-            best_candidate = &candidate;
+                parent_reading_select_result->selected_marks);
+            LOG_DEBUG(logger, "{}", candidate.stat->description);
         }
-        else
+        else if (candidate.stat && candidate.stat->description.empty())
         {
-            LOG_DEBUG(
-                logger,
-                "Projection {} with {} marks is less efficient than current best candidate {} with {} marks",
+            candidate.stat->description = fmt::format(
+                "Projection {} is usable but requires reading {} marks, which is less efficient than projection {} with {} marks",
                 candidate.projection->name,
                 candidate.sum_marks,
                 best_candidate->projection->name,
                 best_candidate->sum_marks);
+            LOG_DEBUG(logger, "{}", candidate.stat->description);
         }
     }
 
-    if (!best_candidate)
-    {
-        reading->setAnalyzedResult(std::move(ordinary_reading_select_result));
-        return {};
-    }
-
     auto storage_snapshot = reading->getStorageSnapshot();
     auto proj_snapshot = std::make_shared<StorageSnapshot>(storage_snapshot->storage, best_candidate->projection->metadata);
     auto query_info_copy = query_info;
-    query_info_copy.prewhere_info = nullptr;
+    if (query.dag && query.filter_node)
+    {
+        query_info_copy.prewhere_info = std::make_shared<PrewhereInfo>();
+        query_info_copy.prewhere_info->need_filter = true;
+        query_info_copy.prewhere_info->remove_prewhere_column = true;
+        query_info_copy.prewhere_info->prewhere_actions = std::move(*query.dag);
+        query_info_copy.prewhere_info->prewhere_column_name = query.filter_node->result_name;
+    }
```

Comment:
I don't fully understand why this is needed. May you please add a comment?

Reply:
Oops. This was meant for a later PR where we’ll use the prewhere expression for projection index reading. I’ll take it out for now.